src/constructs/core/stack.test.ts

```diff
@@ -78,4 +84,41 @@ describe("The GuStack construct", () => {
       "Attempting to read parameter i-do-not-exist which does not exist"
     );
   });
+
+  it("should warn when calling withStageDependentValue with the INFRA stage", () => {
+    const stack = new GuStack(new App(), "Test", { stack: "test" });
+
+    stack.withStageDependentValue({
+      app: "test",
+      variableName: "test",
+      stageValues: {
+        [StageForInfrastructure]: 1,
+      },
+    });
+    expect(warn).toHaveBeenCalledWith(
+      "GuStack does not have a stage of INFRA. Setting a mapping value for it has no impact."
+    );
+  });
+});
+
+describe("The GuStackForInfrastructure construct", () => {
+  it("should have a stage of INFRA", () => {
+    const stack = new GuStackForInfrastructure(new App(), "Test", { stack: "test" });
+    expect(stack.stage).toBe("INFRA");
+  });
+
+  it("should throw when calling withStageDependentValue with a non-INFRA stage", () => {
+    const stack = new GuStackForInfrastructure(new App(), "Test", { stack: "test" });
+
+    expect(() => {
+      stack.withStageDependentValue({
+        app: "test",
+        variableName: "test",
+        stageValues: {
+          [Stage.CODE]: 1,
+          [Stage.PROD]: 2,
+        },
+      });
+    }).toThrowError("Mapping doesn't contain top-level key 'INFRA'");
```

Comment:
This is an error thrown by AWS CDK. We could provide a custom error message, but it feels a little unnecessary as we can rely on the strongly typed `stageValues` instead.